rfcs/0005-collection-context.md

```diff
@@ -0,0 +1,494 @@
+in the graph are directed towards outward scope. For example, an AST node in a function body
+will have an edge pointing outward in scope towards the AST node of the function declaration
+itself. And that function declaration will have and edge pointing outward in scope to the
+file its in. In this way, the root of the graph is the entire application itself, as it
+encompasses the most outward scope.
+
+Modeling scope in this way is useful because it allows the compiler to 'look ahead' at
+'future AST nodes' during type collection, type inference, etc. We can think of it in this way.
+Given this example (it's a silly example, don't think about the fact that this will never
+terminate, that doesn't matter for what we are looking at):
+
+```rust
+fn ping(n: u64) -> u64 {
+    return pong(n);
+}
+
+fn pong(n: u64) -> u64 {
+    return ping(n);
+}
+```
+
+During type inference, when evaluating the AST node for `return pong(n);`, we will need to
+know 1) the type signature of `pong` and 2) if `pong` exists at all and is in scope, but we
+don't know this information because we haven't done type inference on `pong` yet. Because these
+functions are mutually recursive, it is not possible to determine an ordering for which to do
+type inference. The collection context gives the ability for the compiler to either 'look
+forward' or 'look backward' in the AST given any location where type inference is currently 'standing'.
+
+## Mutually Recursive Dependencies
+
+Take these two Sway files (currently does not compile):
+
+`alice.sw`:
+
+```rust
+library alice;
+
+use ::bob::*;
+
+fn alice_fn(n: u64) -> u64 {
+    if n == 0 {
+        return 99;
+    } else {
+        return bob_fn(n-1);
+    }
+}
+```
+
+`bob.sw`:
+
+```rust
+library bob;
+
+use ::alice::*;
+
+fn bob_fn(n: u64) -> u64 {
+        if n == 0 {
+        return 1;
+    } else {
+        return alice_fn(n-1);
+    }
+}
+```
+
+In this case, the collection context would allow the compiler to immutably search the contents of both
+`alice.sw` and `bob.sw`, in any order, regardless of whether `alice_fn` is being type checked first or
+`bob_fn` is.
+
+## "Did you forget to import X in file Y?" Compiler Error
+
+Take the same example as above, but this time the user has forgotten an import statement:
+
+`alice.sw`:
+
+```rust
+library alice;
+
+// use ::bob::*;
+//
+// ^^^ they forgot this
+
+fn alice_fn(n: u64) -> u64 {
+    if n == 0 {
+        return 99;
+    } else {
+        return bob_fn(n-1);
+    }
+}
+```
+
+`bob.sw`:
+
+```rust
+library bob;
+
+use ::alice::*;
+
+fn bob_fn(n: u64) -> u64 {
+        if n == 0 {
+        return 1;
+    } else {
+        return alice_fn(n-1);
+    }
+}
+```
+
+When type checking `alice_fn`, the compiler is able to use the collection context to see that there does exist
+a `bob_fn` function, but that it is out of scope. This information is able to be used to gives better
+error messages and better error recovery.
```

Comment:
Nice! From the LSP side, we could provide a pop-up box asking the user if they would like to auto-import the missing library.